include/exadg/solvers_and_preconditioners/multigrid/multigrid_preconditioner_base.cpp

```diff
@@ -306,6 +306,68 @@ MultigridPreconditionerBase<dim, Number>::check_levels(std::vector<MGLevelInfo>
   }
 }
 
+/**
+ * Similar to dealii::MGTransferGlobalCoarseningTools::create_geometric_coarsening_sequence
+ * with the difference that the (coarse-grid) p:d:T is converted to a p:f:T
+ * right aways.
+ */
+template<int dim, int spacedim>
+std::vector<std::shared_ptr<Triangulation<dim, spacedim> const>>
+create_geometric_coarsening_sequence(Triangulation<dim, spacedim> const & fine_triangulation_in)
+{
+  std::vector<std::shared_ptr<Triangulation<dim, spacedim> const>> coarse_grid_triangulations(
+    fine_triangulation_in.n_global_levels());
+
+  coarse_grid_triangulations.back().reset(&fine_triangulation_in, [](auto *) {
+    // empty deleter, since fine_triangulation_in is an external field
+    // and its destructor is called somewhere else
+  });
+
+  // for a single level nothing has to be done
+  if(fine_triangulation_in.n_global_levels() > 1)
+  {
+    auto const fine_triangulation =
+      dynamic_cast<parallel::distributed::Triangulation<dim, spacedim> const *>(
+        &fine_triangulation_in);
+
+    Assert(fine_triangulation, ExcNotImplemented());
+
+    // clone distributed triangulation for coarsening
+    parallel::distributed::Triangulation<dim, spacedim> temp_tria(
+      fine_triangulation->get_communicator(), fine_triangulation->get_mesh_smoothing());
+
+    temp_tria.copy_triangulation(*fine_triangulation);
+
+    // create coarse meshes
+    for(unsigned int l = (fine_triangulation->n_global_levels() - 1); l > 0; --l)
+    {
+      // coarsen mesh
+      temp_tria.coarsen_global();
+
+      // extract relevant information form distributed triangulation
+      auto const construction_data =
```

Comment:
I would shift this line to lines 360/361 if possible (since `construction_data` does not seem to be used before?).

Reply:
Done.